### PR TITLE
IGNITE-21301 Sync raft log before flush in all storage engines

### PR DESCRIPTION
**Pull request by Phillippko**
https://issues.apache.org/jira/browse/IGNITE-21301

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/Checkpointer.java [hunk 1]
```diff
@@ -340,6 +346,14 @@ void doCheckpoint() throws IgniteInternalCheckedException {
                     }
                 }
 
+                try {
+                    logSyncer.sync();
+                } catch (Exception e) {
+                    log.error("Failed to sync raft log during checkpoint processing ", e);
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/Checkpointer.java:352]: Please don't mention raft

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/logit/LogitLogStorageFactory.java [hunk 2]
```diff
@@ -92,6 +92,11 @@ public void close() {
         ExecutorServiceHelper.shutdownAndAwaitTermination(checkpointExecutor);
     }
 
+    @Override
+    public void sync() {
+        // No-op.
```
Comment (ibessonov) [inline modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/logit/LogitLogStorageFactory.java:97]: Please throw an exception instead. It's not a no-op.

File: modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/flush/RocksDbFlushListener.java [hunk 3]
```diff
@@ -59,6 +59,7 @@ class RocksDbFlushListener extends AbstractEventListener {
     /** {@inheritDoc} */
     @Override
     public void onFlushBegin(RocksDB db, FlushJobInfo flushJobInfo) {
+
```
Comment (ibessonov) [inline modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/flush/RocksDbFlushListener.java:62]: Excuse me?

File: modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/flush/RocksDbFlusher.java [hunk 4]
```diff
@@ -192,6 +192,14 @@ public CompletableFuture<Void> awaitFlush(boolean schedule) {
             future = flushFuturesBySequenceNumber.computeIfAbsent(dbSequenceNumber, s -> new CompletableFuture<>());
         }
 
+        try {
+            db.syncWal();
```
Comment (ibessonov) [inline modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/flush/RocksDbFlusher.java:196]: This is a wrong place

**Review by ibessonov — COMMENTED** [review]
I see no new tests

File: modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/flush/RocksDbFlusher.java (hunk 4 above)
Comment (Phillippko) [reply modules/rocksdb-common/src/main/java/org/apache/ignite/internal/rocksdb/flush/RocksDbFlusher.java]: fixed

File: modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/logit/LogitLogStorageFactory.java (hunk 2 above)
Comment (Phillippko) [reply modules/raft/src/main/java/org/apache/ignite/internal/raft/storage/logit/LogitLogStorageFactory.java]: fixed, added a todo

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/Checkpointer.java (hunk 1 above)
Comment (Phillippko) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/Checkpointer.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/internal/components/LogSyncer.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.components;
+
+/**  Interface to synchronize write-ahead log. Operates only for persistent physical memory log storages. */
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/components/LogSyncer.java:20]:
```suggestion
/** Interface to synchronize write-ahead log. Operates only for persistent physical memory log storages. */
```

Comment (ibessonov) [reply modules/core/src/main/java/org/apache/ignite/internal/components/LogSyncer.java:20]: What's a "physical memory log storages"? Looks a bit confusing.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/Checkpointer.java [hunk 6]
```diff
@@ -340,6 +346,14 @@ void doCheckpoint() throws IgniteInternalCheckedException {
                     }
                 }
 
+                try {
+                    logSyncer.sync();
+                } catch (Exception e) {
+                    log.error("Failed to sync write-ahead log during checkpoint processing ", e);
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/Checkpointer.java:352]:
```suggestion
                    log.error("Failed to sync write-ahead log during checkpoint", e);
```

Comment (ibessonov) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/Checkpointer.java:352]: I don't remember whether we should end the statement with a dot or not, please check other messages

File: modules/rocksdb-common/src/test/java/org/apache/ignite/internal/rocksdb/flush/RocksDbFlushListenerTest.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+                new IgniteSpinBusyLock(),
+                new ScheduledThreadPoolExecutor(20),
+                Executors.newSingleThreadExecutor(),
+                () -> 0,
+                () -> {} // No-op.
+        );
+
+        RocksDbFlushListener listener = new RocksDbFlushListener(flusher);
+
+        listener.onFlushBegin(mockDb, mock(FlushJobInfo.class));
```
Comment (ibessonov) [inline modules/rocksdb-common/src/test/java/org/apache/ignite/internal/rocksdb/flush/RocksDbFlushListenerTest.java:48]: Oh no, that's not what you need to check, ideally. Can we start a table with rocksdb engine, call "flush" and check that the log is synced? That would be much more robust. Also, such a test would be reusable. We can add it to base class with generic storage tests, only enabling it for non-volatile engines, of course (using `assumeTrue`, for example)

File: modules/core/src/main/java/org/apache/ignite/internal/components/LogSyncer.java (hunk 5 above)
Comment (Phillippko) [reply modules/core/src/main/java/org/apache/ignite/internal/components/LogSyncer.java]: removed "physical" bit

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/Checkpointer.java (hunk 6 above)
Comment (Phillippko) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/persistence/checkpoint/Checkpointer.java]: In most messages there is no dot

**Review by ibessonov — APPROVED** [review]

**Pull request merged**